FreeRTOS-Plus/Demo/AWS/Fleet_Provisioning_Windows_Simulator/Fleet_Provisioning_With_CSR_Demo/pkcs11_operations.c

```diff
@@ -0,0 +1,1038 @@
+    CK_RV xPkcs11Ret = CKR_OK;
+    mbedtls_pk_context xPrivKey;
+    mbedtls_pk_info_t xPrivKeyInfo;
+    mbedtls_ecdsa_context xEcdsaContext;
+    mbedtls_x509write_csr xReq;
+    int32_t usMbedtlsRet = -1;
+    const mbedtls_pk_info_t * pxHeader = mbedtls_pk_info_from_type( MBEDTLS_PK_ECKEY );
+
+    assert( pcPrivKeyLabel != NULL );
+    assert( pcPubKeyLabel != NULL );
+    assert( pcCsrBuffer != NULL );
+    assert( pxOutCsrLength != NULL );
+
+    xPkcs11Ret = prvGenerateKeyPairEC( xP11Session,
+                                       pcPrivKeyLabel,
+                                       pcPubKeyLabel,
+                                       &xPrivKeyHandle,
+                                       &xPubKeyHandle );
+
+    if( xPkcs11Ret == CKR_OK )
+    {
+        mbedtls_x509write_csr_init( &xReq );
+        mbedtls_x509write_csr_set_md_alg( &xReq, MBEDTLS_MD_SHA256 );
+
+        usMbedtlsRet = mbedtls_x509write_csr_set_key_usage( &xReq, MBEDTLS_X509_KU_DIGITAL_SIGNATURE );
+
+        if( usMbedtlsRet == 0 )
+        {
+            usMbedtlsRet = mbedtls_x509write_csr_set_ns_cert_type( &xReq, MBEDTLS_X509_NS_CERT_TYPE_SSL_CLIENT );
+        }
+
+        if( usMbedtlsRet == 0 )
+        {
+            usMbedtlsRet = mbedtls_x509write_csr_set_subject_name( &xReq, democonfigCSR_SUBJECT_NAME );
+        }
+
+        if( usMbedtlsRet == 0 )
+        {
+            mbedtls_pk_init( &xPrivKey );
+        }
+
+        if( usMbedtlsRet == 0 )
+        {
+            usMbedtlsRet = prvExtractEcPublicKey( xP11Session, &xEcdsaContext, xPubKeyHandle );
+        }
+
+        if( usMbedtlsRet == 0 )
+        {
+            xSigningContext.p11Session = xP11Session;
+            xSigningContext.p11PrivateKey = xPrivKeyHandle;
+
+            memcpy( &xPrivKeyInfo, pxHeader, sizeof( mbedtls_pk_info_t ) );
+
+            xPrivKeyInfo.sign_func = prvPrivateKeySigningCallback;
+            xPrivKey.pk_info = &xPrivKeyInfo;
+            xPrivKey.pk_ctx = &xEcdsaContext;
+
+            mbedtls_x509write_csr_set_key( &xReq, &xPrivKey );
+
+            usMbedtlsRet = mbedtls_x509write_csr_pem( &xReq, ( unsigned char * ) pcCsrBuffer,
+                                                      xCsrBufferLength, &prvRandomCallback,
+                                                      &xP11Session );
+        }
+
+        mbedtls_x509write_csr_free( &xReq );
+        mbedtls_ecdsa_free( &xEcdsaContext );
+        mbedtls_ecp_group_free( &( xEcdsaContext.grp ) );
+    }
+
+    *pxOutCsrLength = strlen( pcCsrBuffer );
+
+    return( usMbedtlsRet == 0 );
+}
+
+/*-----------------------------------------------------------*/
+
+bool xLoadCertificate( CK_SESSION_HANDLE xP11Session,
+                       const char * pcCertificate,
+                       const char * pcLabel,
+                       size_t xCertificateLength )
+{
+    CK_RV xRet;
+
+    assert( pcCertificate != NULL );
+    assert( pcLabel != NULL );
+
+    xRet = prvProvisionCertificate( xP11Session,
```

Comment:
Is this intermediate function necessary? prvProvisionCertificate is only called from here.

Reply:
Correct - moved prvProvisionCertificate into xLoadCertificate in this commit: https://github.com/FreeRTOS/FreeRTOS/pull/724/commits/ea35474af129619e5f72990a49be1c729527f0b8